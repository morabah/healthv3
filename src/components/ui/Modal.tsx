/**
 * Modal Component
 * An accessible modal dialog component built with Headless UI
 */
import React, { Fragment } from 'react';
import { Dialog, Transition } from '@headlessui/react';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faXmark } from '@fortawesome/free-solid-svg-icons';
import Button from './Button';

export interface ModalProps {
  isOpen: boolean;
  onClose: () => void;
  title?: string;
  children: React.ReactNode;
  size?: 'sm' | 'md' | 'lg' | 'xl' | 'full';
  showCloseButton?: boolean;
  closeOnOverlayClick?: boolean;
  initialFocus?: React.MutableRefObject<HTMLElement | null>;
  footer?: React.ReactNode;
}

const Modal: React.FC<ModalProps> = ({
  isOpen,
  onClose,
  title,
  children,
  size = 'md',
  showCloseButton = true,
  closeOnOverlayClick = true,
  initialFocus,
  footer
}) => {
  // Size classes
  const sizeClasses = {
    sm: 'max-w-sm',
    md: 'max-w-md',
    lg: 'max-w-lg',
    xl: 'max-w-xl',
    full: 'max-w-full mx-4'
  };

  return (
    <Transition appear show={isOpen} as={Fragment}>
      <Dialog 
        as="div" 
        className="relative z-50" 
        onClose={closeOnOverlayClick ? onClose : () => {}}
        initialFocus={initialFocus}
      >
        {/* Background overlay */}
        <Transition.Child
          as={Fragment}
          enter="ease-out duration-300"
          enterFrom="opacity-0"
          enterTo="opacity-100"
          leave="ease-in duration-200"
          leaveFrom="opacity-100"
          leaveTo="opacity-0"
        >
          <div className="fixed inset-0 bg-black bg-opacity-25 backdrop-blur-sm" />
        </Transition.Child>

        <div className="fixed inset-0 overflow-y-auto">
          <div className="flex min-h-full items-center justify-center p-4 text-center">
            <Transition.Child
              as={Fragment}
              enter="ease-out duration-300"
              enterFrom="opacity-0 scale-95"
              enterTo="opacity-100 scale-100"
              leave="ease-in duration-200"
              leaveFrom="opacity-100 scale-100"
              leaveTo="opacity-0 scale-95"
            >
              <Dialog.Panel 
                className={`w-full ${sizeClasses[size]} transform overflow-hidden rounded-lg bg-white p-6 text-left align-middle shadow-xl transition-all`}
              >
                {/* Header */}
                {(title || showCloseButton) && (
                  <div className="flex items-center justify-between mb-4">
                    {title && (
                      <Dialog.Title
                        as="h3"
                        className="text-lg font-medium leading-6 text-gray-900"
                      >
                        {title}
                      </Dialog.Title>
                    )}
                    
                    {showCloseButton && (
                      <button
                        type="button"
                        className="text-gray-400 hover:text-gray-500 focus:outline-none focus:ring-2 focus:ring-blue-500 focus:ring-offset-2 rounded-full p-1"
                        onClick={onClose}
                      >
                        <span className="sr-only">Close</span>
                        <FontAwesomeIcon icon={faXmark} className="h-5 w-5" aria-hidden="true" />
                      </button>
                    )}
                  </div>
                )}
                
                {/* Content */}
                <div className="mt-2">
                  {children}
                </div>
                
                {/* Footer */}
                {footer && (
                  <div className="mt-6 flex justify-end space-x-3">
                    {footer}
                  </div>
                )}
              </Dialog.Panel>
            </Transition.Child>
          </div>
        </div>
      </Dialog>
    </Transition>
  );
};

export default Modal;
