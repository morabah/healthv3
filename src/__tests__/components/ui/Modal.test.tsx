import React from 'react';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import '@testing-library/jest-dom';
import Modal from '@/components/ui/Modal';

// Mock Headless UI's Dialog component to make it work in tests
jest.mock('@headlessui/react', () => {
  const DialogMock = ({ children, ...props }: any) => (
    <div data-testid="dialog" {...props}>
      {children}
    </div>
  );
  
  DialogMock.Panel = ({ children, className }: any) => (
    <div data-testid="dialog-panel" className={className}>
      {children}
    </div>
  );
  
  DialogMock.Title = ({ children, className }: any) => (
    <h3 data-testid="dialog-title" className={className}>
      {children}
    </h3>
  );
  
  const TransitionMock = ({ show, children }: any) => (show ? children : null);
  
  TransitionMock.Child = ({ children }: any) => <>{children}</>;
  
  return {
    Dialog: DialogMock,
    Transition: TransitionMock
  };
});

describe('Modal Component', () => {
  const mockOnClose = jest.fn();

  beforeEach(() => {
    mockOnClose.mockClear();
  });

  it('renders correctly when open', () => {
    render(
      <Modal isOpen={true} onClose={mockOnClose} title="Test Modal">
        <p data-testid="modal-content">Modal content</p>
      </Modal>
    );

    expect(screen.getByTestId('dialog-title')).toHaveTextContent('Test Modal');
    expect(screen.getByTestId('modal-content')).toBeInTheDocument();
  });

  it('does not render when closed', () => {
    render(
      <Modal isOpen={false} onClose={mockOnClose} title="Test Modal">
        <p data-testid="modal-content">Modal content</p>
      </Modal>
    );

    expect(screen.queryByTestId('dialog-title')).not.toBeInTheDocument();
    expect(screen.queryByTestId('modal-content')).not.toBeInTheDocument();
  });

  it('calls onClose when close button is clicked', async () => {
    render(
      <Modal isOpen={true} onClose={mockOnClose} title="Test Modal">
        <p>Modal content</p>
      </Modal>
    );

    const closeButton = screen.getByRole('button', { name: /close/i });
    await userEvent.click(closeButton);
    expect(mockOnClose).toHaveBeenCalledTimes(1);
  });

  it('renders with different sizes', () => {
    const { rerender } = render(
      <Modal isOpen={true} onClose={mockOnClose} size="sm">
        <p>Small modal</p>
      </Modal>
    );

    expect(screen.getByTestId('dialog-panel')).toHaveClass('max-w-sm');

    rerender(
      <Modal isOpen={true} onClose={mockOnClose} size="lg">
        <p>Large modal</p>
      </Modal>
    );

    expect(screen.getByTestId('dialog-panel')).toHaveClass('max-w-lg');

    rerender(
      <Modal isOpen={true} onClose={mockOnClose} size="full">
        <p>Full-width modal</p>
      </Modal>
    );

    expect(screen.getByTestId('dialog-panel')).toHaveClass('max-w-full');
  });

  it('renders without a close button when showCloseButton is false', () => {
    render(
      <Modal isOpen={true} onClose={mockOnClose} showCloseButton={false}>
        <p>Modal without close button</p>
      </Modal>
    );

    expect(screen.queryByRole('button', { name: /close/i })).not.toBeInTheDocument();
  });

  it('renders with a footer when provided', () => {
    render(
      <Modal 
        isOpen={true} 
        onClose={mockOnClose} 
        footer={
          <>
            <button>Cancel</button>
            <button>Submit</button>
          </>
        }
      >
        <p>Modal with footer</p>
      </Modal>
    );

    expect(screen.getByRole('button', { name: /cancel/i })).toBeInTheDocument();
    expect(screen.getByRole('button', { name: /submit/i })).toBeInTheDocument();
  });

  it('matches snapshot', () => {
    const { container } = render(
      <Modal isOpen={true} onClose={mockOnClose} title="Snapshot Test">
        <p>Modal content for snapshot</p>
      </Modal>
    );
    expect(container).toMatchSnapshot();
  });
});
