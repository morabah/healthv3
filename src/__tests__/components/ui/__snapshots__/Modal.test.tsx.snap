// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal Component matches snapshot 1`] = `
<div>
  <div
    as="div"
    class="relative z-50"
    data-testid="dialog"
  >
    <div
      class="fixed inset-0 bg-black bg-opacity-25 backdrop-blur-sm"
    />
    <div
      class="fixed inset-0 overflow-y-auto"
    >
      <div
        class="flex min-h-full items-center justify-center p-4 text-center"
      >
        <div
          class="w-full max-w-md transform overflow-hidden rounded-lg bg-white p-6 text-left align-middle shadow-xl transition-all"
          data-testid="dialog-panel"
        >
          <div
            class="flex items-center justify-between mb-4"
          >
            <h3
              class="text-lg font-medium leading-6 text-gray-900"
              data-testid="dialog-title"
            >
              Snapshot Test
            </h3>
            <button
              class="text-gray-400 hover:text-gray-500 focus:outline-none focus:ring-2 focus:ring-blue-500 focus:ring-offset-2 rounded-full p-1"
              type="button"
            >
              <span
                class="sr-only"
              >
                Close
              </span>
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-xmark h-5 w-5"
                data-icon="xmark"
                data-prefix="fas"
                focusable="false"
                role="img"
                viewBox="0 0 384 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M342.6 150.6c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L192 210.7 86.6 105.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3L146.7 256 41.4 361.4c-12.5 12.5-12.5 32.8 0 45.3s32.8 12.5 45.3 0L192 301.3 297.4 406.6c12.5 12.5 32.8 12.5 45.3 0s12.5-32.8 0-45.3L237.3 256 342.6 150.6z"
                  fill="currentColor"
                />
              </svg>
            </button>
          </div>
          <div
            class="mt-2"
          >
            <p>
              Modal content for snapshot
            </p>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
